// @cloudflare/component-dropdown/src/DropdownLink.js
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from '@cloudflare/component-link';
import { createStyledComponent } from '@cloudflare/style-container';

import DropdownRegistry from './DropdownRegistry';

const styled = ({ disabled, theme }) => ({
  position: 'relative',
  zIndex: 1,

  '&:first-child': {
    borderTopWidth: 0
  },

  '& > a': {
    display: 'block',

    paddingTop: theme.space[1],
    paddingBottom: theme.space[1],
    paddingLeft: theme.space[3],
    paddingRight: theme.space[3],

    cursor: disabled ? 'default' : 'pointer',
    whiteSpace: 'nowrap',

    '&:hover': {
      color: disabled ? theme.colors.gray[4] : theme.colors.white,
      background: disabled ? theme.colors.white : theme.colors.blue[4]
    },

    '&:focus': {
      color: disabled ? theme.colors.gray[4] : theme.colors.white,
      background: disabled ? theme.colors.white : theme.colors.blue[6],
      outline: '5px auto -webkit-focus-ring-color',
      outlineOffset: '-1px'
    },

    '&:active': {
      color: theme.colors.white,
      background: theme.colors.blue[4]
    }
  }
});

class DropdownLink extends React.Component {
  constructor(props, context) {
    if (!props.to && !props.onClick) {
      throw new Error(
        '<DropdownLink/> requires either a `to` or `onClick` prop'
      );
    }

    super(props, context);
    this.dropdownRegistry = context.dropdownRegistry;

    this.handleLinkBlur = this.handleLinkBlur.bind(this);
    this.handleLinkFocus = this.handleLinkFocus.bind(this);
    this.focus = this.focus.bind(this);
  }

  componentDidMount() {
    this.dropdownRegistry.addChild(this);
  }

  componentWillUnmount() {
    this.dropdownRegistry.removeChild(this);
  }

  focus() {
    this.link.focus();
  }

  handleLinkFocus() {
    this.dropdownRegistry.setFocusedChild(this);
  }

  handleLinkBlur() {
    this.dropdownRegistry.removeFocusedChild();
  }

  render() {
    return (
      <li className={this.props.className} role="menuitem">
        <Link
          innerRef={node => (this.link = node)}
          to={this.props.to}
          onClick={this.props.onClick}
          onFocus={this.handleLinkFocus}
          onBlur={this.handleLinkBlur}
          disabled={this.props.disabled}
        >
          {this.props.children}
        </Link>
      </li>
    );
  }
}

DropdownLink.propTypes = {
  disabled: PropTypes.bool,
  className: PropTypes.string,
  to: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node
};

DropdownLink.contextTypes = {
  dropdownRegistry: PropTypes.instanceOf(DropdownRegistry).isRequired
};

DropdownLink.displayName = 'DropdownLink';
export default createStyledComponent(styled, DropdownLink);
