// @cloudflare/types/es/api/index.js
import { eg } from '@cloudflare/util-en-garde';
export * from './account';
export * from './alert';
export * from './apiKey';
export * from './apiToken';
export * from './billing';
export * from './dns';
export * from './domain';
export * from './domainResourceGroup';
export * from './zoneVersioning';
export * from './captcha';
export * from './entitlement';
export * from './flags';
export * from './graphQLSettings';
export * from './lb';
export * from './marketing';
export * from './membership';
export * from './permissions';
export * from './organization';
export * from './pageRule';
export * from './pages';
export * from './policy';
export * from './profile';
export * from './ruleset';
export * from './subscription';
export * from './two-factor';
export * from './user';
export * from './zone';
export * from './ratePlan';
export * from './settings';
export * from './spectrum';
export * from './zendesk';
export * from './railgun';
export * from './oAuthConsentApp';
export const ApiResponseBody = eg.object({
  result: eg.any,
  success: eg.boolean,
  errors: eg.array(eg.object({
    code: eg.number.optional,
    message: eg.string
  })),
  messages: eg.array(eg.any).optional,
  result_info: eg.any.optional
});
