// @cloudflare/component-page/src/PageHeader.js
import React from 'react';
import PropTypes from 'prop-types';
import { createComponent } from '@cloudflare/style-container';
import { Label } from '@cloudflare/component-label';

const PageHeading = createComponent(
  ({ theme }) => ({
    fontSize: theme.fontSizes[6],
    marginTop: theme.space[0],
    marginBottom: theme.space[0],
    lineHeight: 1.2,
    fontWeight: 400
  }),
  'h1'
);
PageHeading.setDisplayName('PageHeading');

const PageSubheading = createComponent(
  ({ theme }) => ({
    fontSize: theme.fontSizes[5],
    marginTop: theme.space[2],
    marginBottom: 0,
    lineHeight: 1.2,
    color: theme.colors.gray[3],
    fontWeight: theme.fontWeightLight
  }),
  'h2'
);
PageSubheading.setDisplayName('PageSubheading');

const Beta = createComponent(
  ({ theme }) => ({
    marginLeft: theme.space[2],
    verticalAlign: 'middle'
  }),
  Label
);
Beta.setDisplayName('Beta');

const PageHeader = ({ title, subtitle, beta, className }) => {
  return (
    <header className={className}>
      <PageHeading>
        {title}
        {beta && <Beta type="warning">Beta</Beta>}
      </PageHeading>
      {subtitle && <PageSubheading>{subtitle}</PageSubheading>}
    </header>
  );
};

PageHeader.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  beta: PropTypes.bool
};

export default createComponent(() => {}, PageHeader);
