// @cloudflare/component-page/lib/PageContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styles = function styles(_ref) {
  var type = _ref.type;
  return {
    marginLeft: 'auto',
    marginRight: 'auto',
    width: '90%',
    maxWidth: type === 'narrow' ? '64em' : '79em'
  };
};

var PageContent = function PageContent(_ref2) {
  var className = _ref2.className,
      children = _ref2.children;
  return _react.default.createElement("div", {
    className: className
  }, children);
};

PageContent.propTypes = {
  className: _propTypes.default.string,
  children: _propTypes.default.node,
  type: _propTypes.default.oneOf(['narrow', 'wide'])
};
PageContent.defaultProps = {
  type: 'wide'
};
PageContent.displayName = 'PageContent';

var _default = (0, _styleContainer.createComponent)(styles, PageContent);

exports.default = _default;
