// @cloudflare/component-page/lib/PageHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

var _componentLabel = require("@cloudflare/component-label");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PageHeading = (0, _styleContainer.createComponent)(function (_ref) {
  var theme = _ref.theme;
  return {
    fontSize: theme.fontSizes[6],
    marginTop: theme.space[0],
    marginBottom: theme.space[0],
    lineHeight: 1.2,
    fontWeight: 400
  };
}, 'h1');
PageHeading.setDisplayName('PageHeading');
var PageSubheading = (0, _styleContainer.createComponent)(function (_ref2) {
  var theme = _ref2.theme;
  return {
    fontSize: theme.fontSizes[5],
    marginTop: theme.space[2],
    marginBottom: 0,
    lineHeight: 1.2,
    color: theme.colors.gray[3],
    fontWeight: theme.fontWeightLight
  };
}, 'h2');
PageSubheading.setDisplayName('PageSubheading');
var Beta = (0, _styleContainer.createComponent)(function (_ref3) {
  var theme = _ref3.theme;
  return {
    marginLeft: theme.space[2],
    verticalAlign: 'middle'
  };
}, _componentLabel.Label);
Beta.setDisplayName('Beta');

var PageHeader = function PageHeader(_ref4) {
  var title = _ref4.title,
      subtitle = _ref4.subtitle,
      beta = _ref4.beta,
      className = _ref4.className;
  return _react.default.createElement("header", {
    className: className
  }, _react.default.createElement(PageHeading, null, title, beta && _react.default.createElement(Beta, {
    type: "warning"
  }, "Beta")), subtitle && _react.default.createElement(PageSubheading, null, subtitle));
};

PageHeader.propTypes = {
  className: _propTypes.default.string,
  title: _propTypes.default.string.isRequired,
  subtitle: _propTypes.default.string,
  beta: _propTypes.default.bool
};

var _default = (0, _styleContainer.createComponent)(function () {}, PageHeader);

exports.default = _default;
