// @cloudflare/component-page/es/PageHeader.js
import React from 'react';
import PropTypes from 'prop-types';
import { createComponent } from '@cloudflare/style-container';
import { Label } from '@cloudflare/component-label';
var PageHeading = createComponent(function (_ref) {
  var theme = _ref.theme;
  return {
    fontSize: theme.fontSizes[6],
    marginTop: theme.space[0],
    marginBottom: theme.space[0],
    lineHeight: 1.2,
    fontWeight: 400
  };
}, 'h1');
PageHeading.setDisplayName('PageHeading');
var PageSubheading = createComponent(function (_ref2) {
  var theme = _ref2.theme;
  return {
    fontSize: theme.fontSizes[5],
    marginTop: theme.space[2],
    marginBottom: 0,
    lineHeight: 1.2,
    color: theme.colors.gray[3],
    fontWeight: theme.fontWeightLight
  };
}, 'h2');
PageSubheading.setDisplayName('PageSubheading');
var Beta = createComponent(function (_ref3) {
  var theme = _ref3.theme;
  return {
    marginLeft: theme.space[2],
    verticalAlign: 'middle'
  };
}, Label);
Beta.setDisplayName('Beta');

var PageHeader = function PageHeader(_ref4) {
  var title = _ref4.title,
      subtitle = _ref4.subtitle,
      beta = _ref4.beta,
      className = _ref4.className;
  return React.createElement("header", {
    className: className
  }, React.createElement(PageHeading, null, title, beta && React.createElement(Beta, {
    type: "warning"
  }, "Beta")), subtitle && React.createElement(PageSubheading, null, subtitle));
};

PageHeader.propTypes = {
  className: PropTypes.string,
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  beta: PropTypes.bool
};
export default createComponent(function () {}, PageHeader);
