// @cloudflare/intl-types/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SupportedLocales = exports.Interpolation = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Interpolation = /*#__PURE__*/function () {
  function Interpolation(value, keys) {
    _classCallCheck(this, Interpolation);

    _defineProperty(this, "value", void 0);

    _defineProperty(this, "keys", void 0);

    this.value = value;
    this.keys = keys;
  }

  _createClass(Interpolation, [{
    key: "toString",
    value: function toString() {
      return this.value;
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      return this.value;
    }
  }]);

  return Interpolation;
}();

exports.Interpolation = Interpolation;
var SupportedLocales; // A Locale is the string actually used in the frontend, like 'en-US'
// This is aliased to make a slightly easier to read type

exports.SupportedLocales = SupportedLocales;

(function (SupportedLocales) {
  SupportedLocales["en_US"] = "en-US";
  SupportedLocales["de_DE"] = "de-DE";
  SupportedLocales["zh_CN"] = "zh-CN";
  SupportedLocales["fr_FR"] = "fr-FR";
  SupportedLocales["ja_JP"] = "ja-JP";
  SupportedLocales["es_ES"] = "es-ES";
  SupportedLocales["es_CL"] = "es-CL";
  SupportedLocales["es_EC"] = "es-EC";
  SupportedLocales["es_MX"] = "es-MX";
  SupportedLocales["es_PE"] = "es-PE";
  SupportedLocales["it_IT"] = "it-IT";
  SupportedLocales["ko_KR"] = "ko-KR";
  SupportedLocales["pt_BR"] = "pt-BR";
  SupportedLocales["zh_TW"] = "zh-TW";
})(SupportedLocales || (exports.SupportedLocales = SupportedLocales = {}));
