// @cloudflare/component-heading/lib/Heading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styles = function styles(_ref) {
  var theme = _ref.theme,
      size = _ref.size;
  return {
    // Start at 32px and descend the scale
    fontSize: 7 - size,
    fontWeight: 700,
    lineHeight: theme.lineHeights.title,
    mt: 0,
    p: 0
  };
};

var Heading = function Heading(_ref2) {
  var size = _ref2.size,
      className = _ref2.className,
      children = _ref2.children;
  var tagName = 'h' + size;
  return /*#__PURE__*/_react.default.createElement(tagName, {
    className: className
  }, children);
};

Heading.propTypes = {
  size: _propTypes.default.oneOf([1, 2, 3, 4, 5, 6]).isRequired,
  className: _propTypes.default.string.isRequired,
  children: _propTypes.default.node
};
Heading.displayName = 'Heading';

var _default = (0, _styleContainer.createStyledComponent)(styles, Heading);

exports.default = _default;
