// @cloudflare/component-header/lib/Hamburger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _componentIcon = require("@cloudflare/component-icon");

var _styleContainer = require("@cloudflare/style-container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Wrapper = (0, _styleContainer.createComponent)(function (_ref) {
  var theme = _ref.theme;
  return {
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    borderRight: "1px solid ".concat(theme.colors.gray[7]),
    padding: theme.space[3],
    '-webkit-tap-highlight-color': 'rgba(0, 0, 0, 0)',
    userSelect: 'none',
    cursor: 'pointer',
    tabletWide: {
      display: 'none'
    }
  };
}, 'div', ['onClick']);
Wrapper.setDisplayName('Wrapper');

var Hamburger = function Hamburger(_ref2) {
  var onClick = _ref2.onClick,
      label = _ref2.label;
  return _react.default.createElement(Wrapper, {
    onClick: onClick
  }, _react.default.createElement(_componentIcon.Icon, {
    label: label,
    size: "1.5x",
    type: "hamburger"
  }));
};

Hamburger.propTypes = {
  label: _propTypes.default.string,
  onClick: _propTypes.default.func.isRequired
};
Hamburger.defaultProps = {
  label: 'Open the menu'
};
var _default = Hamburger;
exports.default = _default;
