// @cloudflare/component-header/lib/Header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Header = (0, _styleContainer.createComponent)(function (_ref) {
  var theme = _ref.theme,
      sticky = _ref.sticky;
  return {
    backgroundColor: theme.colors.white,
    display: 'flex',
    justifyContent: 'space-between',
    color: theme.colors.gray[1],
    height: 63,
    top: 0,
    left: 0,
    position: sticky ? 'fixed' : 'static',
    width: '100%',
    borderTop: "3px solid ".concat(theme.colors.cfOrange),
    borderBottom: "1px solid ".concat(theme.colors.gray[8]),
    zIndex: 1200
  };
}, 'header');
Header.setDisplayName('Header');
Header.propTypes = {
  children: _propTypes.default.node.isRequired,
  sticky: _propTypes.default.bool
};
var _default = Header;
exports.default = _default;
