// @cloudflare/types/lib/api/alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvailableAlert = exports.AlertPolicyAPIV2 = exports.AlertDeliveryMechanism = exports.AlertPolicy = exports.EmailParameters = exports.AlertEligibleIntegrations = exports.AlertIntegration = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

// NotificationMechanism is specifies how a notification is delivered -
// via email, SMS, PagerDuty, etc
// type INotificationMechanism = 'email' | 'sms' | 'pagerduty';
var AlertIntegration = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  createdAt: _utilEnGarde.eg.string
});
/**
 * A connection request to Pager Duty that is created before we attempt to link
 * PagerDuty to this account
 */


exports.AlertIntegration = AlertIntegration;

var PagerDutyIntegrationRequest = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string
});

var AlertWebhookType = _utilEnGarde.eg.union([_utilEnGarde.eg.literal('slack'), _utilEnGarde.eg.literal('gchat'), _utilEnGarde.eg.literal('discord'), _utilEnGarde.eg.literal('generic'), _utilEnGarde.eg.literal('splunk'), _utilEnGarde.eg.literal('datadog'), _utilEnGarde.eg.literal('opsgenie'), _utilEnGarde.eg.literal('feishu')]);

var AlertWebhookIntegration = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  url: _utilEnGarde.eg.string,
  createdAt: _utilEnGarde.eg.string,
  secret: _utilEnGarde.eg.string.optional,
  last_success: _utilEnGarde.eg.string.optional,
  last_failure: _utilEnGarde.eg.string.optional
});

var AlertEligibleIntegrations = _utilEnGarde.eg.record(_utilEnGarde.eg.string, _utilEnGarde.eg.object({
  eligible: _utilEnGarde.eg.boolean,
  ready: _utilEnGarde.eg.boolean,
  enabled: _utilEnGarde.eg.boolean,
  type: _utilEnGarde.eg.string
}));

exports.AlertEligibleIntegrations = AlertEligibleIntegrations;

var EmailParameters = _utilEnGarde.eg.object({
  recipients: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
});

exports.EmailParameters = EmailParameters;

// ConditionFilter is pretty much just a condition.
var ConditionFilter = _utilEnGarde.eg.object({
  '>=': _utilEnGarde.eg.array(_utilEnGarde.eg.any)
});

// Type AlertPolicy is the type that represents an alert policy
var AlertPolicy = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string.optional,
  name: _utilEnGarde.eg.string,
  description: _utilEnGarde.eg.string.optional,
  enabled: _utilEnGarde.eg.boolean,
  emails: EmailParameters,
  pager_duty: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional,
  webhooks: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional,
  alert_type: _utilEnGarde.eg.string,
  conditions: ConditionFilter.optional,
  filters: _utilEnGarde.eg.record(_utilEnGarde.eg.string, _utilEnGarde.eg.union([_utilEnGarde.eg.array(_utilEnGarde.eg.string), _utilEnGarde.eg.null])).optional
});

exports.AlertPolicy = AlertPolicy;

var AlertDeliveryMechanism = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string.optional
});

exports.AlertDeliveryMechanism = AlertDeliveryMechanism;

var AlertMechanismMap = _utilEnGarde.eg.record(_utilEnGarde.eg.string, _utilEnGarde.eg.array(AlertDeliveryMechanism));

var AlertPolicyAPIV2 = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string.optional,
  // UUID, nullable for creating new API's
  name: _utilEnGarde.eg.string,
  description: _utilEnGarde.eg.string.optional,
  enabled: _utilEnGarde.eg.boolean,
  conditions: ConditionFilter.optional,
  mechanisms: AlertMechanismMap,
  filters: _utilEnGarde.eg.record(_utilEnGarde.eg.string, _utilEnGarde.eg.union([_utilEnGarde.eg.array(_utilEnGarde.eg.string), _utilEnGarde.eg.null])).optional,
  alert_type: _utilEnGarde.eg.string
});

exports.AlertPolicyAPIV2 = AlertPolicyAPIV2;

var AvailableAlert = _utilEnGarde.eg.object({
  display_name: _utilEnGarde.eg.string,
  type: _utilEnGarde.eg.string,
  description: _utilEnGarde.eg.string.optional
});

exports.AvailableAlert = AvailableAlert;
