// @cloudflare/types/lib/api/entitlement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Entitlement = exports.Feature = exports.Allocation = exports.Interval = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var Interval = _utilEnGarde.eg.object({
  min: _utilEnGarde.eg.number,
  max: _utilEnGarde.eg.number
});

exports.Interval = Interval;

var Allocation = _utilEnGarde.eg.object({
  type: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('bool'), _utilEnGarde.eg.literal('range'), _utilEnGarde.eg.literal('max_count'), _utilEnGarde.eg.literal('enum_number')]),
  value: _utilEnGarde.eg.union([_utilEnGarde.eg.boolean, _utilEnGarde.eg.number, Interval, _utilEnGarde.eg.array(_utilEnGarde.eg.number)])
});

exports.Allocation = Allocation;

var Feature = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.number,
  key: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  feature_set: _utilEnGarde.eg.string
});

exports.Feature = Feature;

var Entitlement = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  created_date: _utilEnGarde.eg.string,
  deleted_date: _utilEnGarde.eg.union([_utilEnGarde.eg.string, _utilEnGarde.eg.null]),
  edited_date: _utilEnGarde.eg.string,
  feature: Feature.optional,
  allocation: Allocation
});

exports.Entitlement = Entitlement;
