// @cloudflare/types/lib/api/graphQLSettings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GraphQLSettings = exports.GraphQLSetting = exports.GraphQLNodeName = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var GraphQLNodeName = _utilEnGarde.eg.union([_utilEnGarde.eg.literal('firewallEventsAdaptive'), _utilEnGarde.eg.literal('firewallEventsAdaptiveByTimeGroups'), _utilEnGarde.eg.literal('firewallEventsAdaptiveGroups'), _utilEnGarde.eg.literal('httpRequestsAdaptiveGroups'), _utilEnGarde.eg.literal('imageResizingRequests1mGroups'), _utilEnGarde.eg.literal('loadBalancingRequestsAdaptive'), _utilEnGarde.eg.literal('loadBalancingRequestsAdaptiveGroups'), _utilEnGarde.eg.literal('healthCheckEventsAdaptive'), _utilEnGarde.eg.literal('healthCheckEventsAdaptiveGroups'), _utilEnGarde.eg.literal('nelReportsAdaptiveGroups'), _utilEnGarde.eg.literal('waitingRoomAnalyticsAdaptiveGroups')]);

exports.GraphQLNodeName = GraphQLNodeName;

var GraphQLSetting = _utilEnGarde.eg.object({
  enabled: _utilEnGarde.eg.boolean,
  maxDuration: _utilEnGarde.eg.number,
  maxNumberOfFields: _utilEnGarde.eg.number.optional,
  maxPageSize: _utilEnGarde.eg.number,
  notOlderThan: _utilEnGarde.eg.number,
  availableFields: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional
});

exports.GraphQLSetting = GraphQLSetting;

var GraphQLSettings = _utilEnGarde.eg.record(GraphQLNodeName, GraphQLSetting);

exports.GraphQLSettings = GraphQLSettings;
