// @cloudflare/types/lib/api/pageRule.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PageRulesQuota = exports.PageRulesSetting = exports.PageRuleProperty = exports.PageRulesSettingForwardingUrl = exports.PageRulesSettingStatusCode = exports.PageRulesSettingCheckboxes = exports.PageRulesSettingString = exports.PageRulesSettingSelect = exports.PageRulesSettingRange = exports.PageRulesSettingToggle = exports.PageRule = exports.PageRuleTarget = exports.PageRulesConstraint = exports.PageRuleAction = exports.PageRuleValue = exports.PageRuleForwardingUrlValue = exports.PageRuleMinifyValue = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var PageRuleMinifyValue = _utilEnGarde.eg.object({
  html: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('on'), _utilEnGarde.eg.literal('off')]).optional,
  css: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('on'), _utilEnGarde.eg.literal('off')]).optional,
  js: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('on'), _utilEnGarde.eg.literal('off')]).optional
});

exports.PageRuleMinifyValue = PageRuleMinifyValue;

var PageRuleForwardingUrlValue = _utilEnGarde.eg.object({
  status_code: _utilEnGarde.eg.number,
  url: _utilEnGarde.eg.string
});

exports.PageRuleForwardingUrlValue = PageRuleForwardingUrlValue;

var PageRuleCacheKeyFieldsValue = _utilEnGarde.eg.object({
  query_string: _utilEnGarde.eg.union([_utilEnGarde.eg.object({
    include: _utilEnGarde.eg.literal('*'),
    exclude: _utilEnGarde.eg.emptyArray
  }), _utilEnGarde.eg.object({
    include: _utilEnGarde.eg.emptyArray,
    exclude: _utilEnGarde.eg.literal('*')
  }), _utilEnGarde.eg.object({
    include: _utilEnGarde.eg.array(_utilEnGarde.eg.string),
    exclude: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
  })]),
  header: _utilEnGarde.eg.object({
    include: _utilEnGarde.eg.array(_utilEnGarde.eg.string),
    exclude: _utilEnGarde.eg.array(_utilEnGarde.eg.literal('origin')),
    check_presence: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
  }),
  cookie: _utilEnGarde.eg.object({
    include: _utilEnGarde.eg.array(_utilEnGarde.eg.string),
    check_presence: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
  }),
  host: _utilEnGarde.eg.object({
    resolved: _utilEnGarde.eg.boolean
  }),
  user: _utilEnGarde.eg.object({
    device_type: _utilEnGarde.eg.boolean,
    geo: _utilEnGarde.eg.boolean,
    lang: _utilEnGarde.eg.boolean
  })
});

var PageRuleValue = _utilEnGarde.eg.union([_utilEnGarde.eg.literal('on'), _utilEnGarde.eg.literal('off'), _utilEnGarde.eg.string, _utilEnGarde.eg.number, PageRuleMinifyValue, PageRuleForwardingUrlValue, PageRuleCacheKeyFieldsValue]);

exports.PageRuleValue = PageRuleValue;

var PageRuleAction = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  value: PageRuleValue.optional
});

exports.PageRuleAction = PageRuleAction;

var PageRulesConstraint = _utilEnGarde.eg.object({
  operator: _utilEnGarde.eg.literal('matches'),
  value: _utilEnGarde.eg.string.optional
});

exports.PageRulesConstraint = PageRulesConstraint;

var PageRuleTarget = _utilEnGarde.eg.object({
  target: _utilEnGarde.eg.literal('url'),
  constraint: PageRulesConstraint
});

exports.PageRuleTarget = PageRuleTarget;

var PageRule = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  priority: _utilEnGarde.eg.number,
  status: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('active'), _utilEnGarde.eg.literal('disabled')]),
  actions: _utilEnGarde.eg.array(PageRuleAction),
  targets: _utilEnGarde.eg.array(PageRuleTarget),
  created_on: _utilEnGarde.eg.string.optional,
  modified_on: _utilEnGarde.eg.string.optional
});

exports.PageRule = PageRule;

var PageRulesSettingToggle = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('value'),
  type: _utilEnGarde.eg.literal('toggle')
});

exports.PageRulesSettingToggle = PageRulesSettingToggle;

var PageRulesSettingRange = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('value'),
  type: _utilEnGarde.eg.literal('range'),
  min: _utilEnGarde.eg.number,
  max: _utilEnGarde.eg.number,
  suggested_values: _utilEnGarde.eg.array(_utilEnGarde.eg.number)
});

exports.PageRulesSettingRange = PageRulesSettingRange;

var PageRulesSettingSelect = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('value'),
  type: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('choice'), _utilEnGarde.eg.literal('select')]),
  multiple: _utilEnGarde.eg.boolean,
  allowEmpty: _utilEnGarde.eg.boolean.optional,
  choices: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
});

exports.PageRulesSettingSelect = PageRulesSettingSelect;

var PageRulesSettingString = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('value'),
  type: _utilEnGarde.eg.literal('string'),
  readonly: _utilEnGarde.eg.boolean
});

exports.PageRulesSettingString = PageRulesSettingString;

var PageRulesSettingCheckboxes = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('value'),
  type: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('choice'), _utilEnGarde.eg.literal('select')]),
  multiple: _utilEnGarde.eg.boolean,
  allowEmpty: _utilEnGarde.eg.boolean.optional,
  choices: _utilEnGarde.eg.array(_utilEnGarde.eg.string)
});

exports.PageRulesSettingCheckboxes = PageRulesSettingCheckboxes;

var PageRulesSettingStatusCode = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('status_code'),
  type: _utilEnGarde.eg.literal('choice'),
  multiple: _utilEnGarde.eg.boolean,
  choices: _utilEnGarde.eg.array(_utilEnGarde.eg.number)
});

exports.PageRulesSettingStatusCode = PageRulesSettingStatusCode;

var PageRulesSettingForwardingUrl = _utilEnGarde.eg.object({
  name: _utilEnGarde.eg.literal('url'),
  type: _utilEnGarde.eg.literal('forwardingUrl')
});

exports.PageRulesSettingForwardingUrl = PageRulesSettingForwardingUrl;

var PageRuleProperty = _utilEnGarde.eg.union([PageRulesSettingToggle, PageRulesSettingRange, PageRulesSettingStatusCode, PageRulesSettingString, PageRulesSettingForwardingUrl, PageRulesSettingSelect, PageRulesSettingCheckboxes]);

exports.PageRuleProperty = PageRuleProperty;

var PageRulesSetting = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  properties: _utilEnGarde.eg.array(PageRuleProperty),
  readonly: _utilEnGarde.eg.boolean.optional
});

exports.PageRulesSetting = PageRulesSetting;

var PageRulesQuota = _utilEnGarde.eg.object({
  quota: _utilEnGarde.eg.number
});

exports.PageRulesQuota = PageRulesQuota;
