// @cloudflare/types/lib/api/permissions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Can = exports.PermKey = exports.Permissions = exports.PermValue = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var PermValue = _utilEnGarde.eg.object({
  read: _utilEnGarde.eg.boolean,
  edit: _utilEnGarde.eg.boolean
});

exports.PermValue = PermValue;

var Permissions = _utilEnGarde.eg.object({
  access: PermValue,
  analytics: PermValue,
  app: PermValue,
  auditlogs: PermValue,
  billing: PermValue,
  blocks: PermValue,
  cache_purge: PermValue,
  dns_records: PermValue,
  lb: PermValue,
  legal: PermValue,
  logs: PermValue,
  member: PermValue,
  organization: PermValue,
  ssl: PermValue,
  stream: PermValue,
  subscription: PermValue,
  waf: PermValue,
  waitingroom: PermValue,
  webhooks: PermValue,
  worker: PermValue,
  zaraz: PermValue,
  zone: PermValue,
  zone_settings: PermValue,
  http_applications: PermValue,
  zone_versioning: PermValue
});

exports.Permissions = Permissions;

var PermKey = _utilEnGarde.eg.keyOf(Permissions);

exports.PermKey = PermKey;

var Can = _utilEnGarde.eg.object({
  read: _utilEnGarde.eg.boolean,
  update: _utilEnGarde.eg.boolean,
  create: _utilEnGarde.eg.boolean,
  delete: _utilEnGarde.eg.boolean,
  list: _utilEnGarde.eg.boolean,
  sign: _utilEnGarde.eg.boolean,
  refresh: _utilEnGarde.eg.boolean.optional
});

exports.Can = Can;
