// @cloudflare/types/lib/api/policy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MembershipsPolicy = exports.PolicyResourceGroup = exports.MembershipsNestedScope = exports.PolicyPermissionGroup = exports.MembershipsPermission = exports.ZoneLevelAccessPolicy = exports.ZoneLevelAccessPolicyScope = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var ZoneLevelAccessPolicyScope = _utilEnGarde.eg.object({
  accountId: _utilEnGarde.eg.string.optional,
  effect: _utilEnGarde.eg.string,
  mode: _utilEnGarde.eg.string,
  zoneId: _utilEnGarde.eg.string.optional,
  resourceGroupId: _utilEnGarde.eg.string.optional
});

exports.ZoneLevelAccessPolicyScope = ZoneLevelAccessPolicyScope;

var ZoneLevelAccessPolicy = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  access: _utilEnGarde.eg.boolean,
  owner: _utilEnGarde.eg.string,
  emails: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional,
  auto_accept: _utilEnGarde.eg.boolean.optional,
  resource_groups: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional,
  permission_groups: _utilEnGarde.eg.array(_utilEnGarde.eg.string),
  scopes: _utilEnGarde.eg.array(ZoneLevelAccessPolicyScope)
});

exports.ZoneLevelAccessPolicy = ZoneLevelAccessPolicy;

var MembershipsPermission = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  key: _utilEnGarde.eg.string,
  implies: _utilEnGarde.eg.array(_utilEnGarde.eg.object({
    id: _utilEnGarde.eg.string,
    key: _utilEnGarde.eg.string
  })).optional
});

exports.MembershipsPermission = MembershipsPermission;

var PolicyPermissionGroup = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  meta: _utilEnGarde.eg.object({
    description: _utilEnGarde.eg.string,
    editable: _utilEnGarde.eg.boolean,
    label: _utilEnGarde.eg.string,
    scopes: _utilEnGarde.eg.string
  }),
  permissions: _utilEnGarde.eg.array(MembershipsPermission).optional
});

exports.PolicyPermissionGroup = PolicyPermissionGroup;

var MembershipsScope = _utilEnGarde.eg.object({
  key: _utilEnGarde.eg.string,
  objects: _utilEnGarde.eg.array(_utilEnGarde.eg.object({
    key: _utilEnGarde.eg.string
  })).optional
});

var MembershipsNestedScope = _utilEnGarde.eg.object({
  key: _utilEnGarde.eg.string,
  objects: _utilEnGarde.eg.array(_utilEnGarde.eg.object({
    key: _utilEnGarde.eg.string
  })),
  subset_of: _utilEnGarde.eg.array(MembershipsScope).optional
});

exports.MembershipsNestedScope = MembershipsNestedScope;

var PolicyResourceGroup = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  meta: _utilEnGarde.eg.object({
    editable: _utilEnGarde.eg.string
  }),
  name: _utilEnGarde.eg.string,
  scope: MembershipsNestedScope
});

exports.PolicyResourceGroup = PolicyResourceGroup;

var MembershipsPolicy = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  access: _utilEnGarde.eg.string,
  permission_groups: _utilEnGarde.eg.array(PolicyPermissionGroup),
  resource_groups: _utilEnGarde.eg.array(PolicyResourceGroup)
});

exports.MembershipsPolicy = MembershipsPolicy;
