// @cloudflare/types/lib/api/spectrum.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpectrumApp = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var _dns = require("./dns");

var SpectrumApp = _utilEnGarde.eg.object({
  dns: _utilEnGarde.eg.object({
    name: _dns.DNSRecord.props.name,
    type: _dns.DNSRecord.props.type
  }).optional,
  edge_ips: _utilEnGarde.eg.object({
    connectivity: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('all'), _utilEnGarde.eg.string]).optional,
    type: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('dynamic'), _utilEnGarde.eg.string])
  }).optional,
  id: _utilEnGarde.eg.string,
  ip_firewall: _utilEnGarde.eg.boolean.optional,
  modified_on: _utilEnGarde.eg.string,
  origin_direct: _utilEnGarde.eg.array(_utilEnGarde.eg.string).optional,
  protocol: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('tcp/34'), _utilEnGarde.eg.string]).optional,
  proxy_protocol: _utilEnGarde.eg.boolean.optional,
  spp: _utilEnGarde.eg.boolean.optional,
  tls: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('on'), _utilEnGarde.eg.literal('off')]).optional,
  traffic_type: _utilEnGarde.eg.union([_utilEnGarde.eg.literal('direct'), _utilEnGarde.eg.string]).optional
});

exports.SpectrumApp = SpectrumApp;
