// @cloudflare/types/lib/api/zoneVersioning.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZoneInfo = void 0;

var _utilEnGarde = require("@cloudflare/util-en-garde");

var HTTPApplication = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  status: _utilEnGarde.eg.string,
  current_version: _utilEnGarde.eg.string,
  last_updated: _utilEnGarde.eg.string
});

var ZoneApplication = _utilEnGarde.eg.object({
  id: _utilEnGarde.eg.string,
  name: _utilEnGarde.eg.string,
  status: _utilEnGarde.eg.string,
  current_version: _utilEnGarde.eg.string,
  last_updated: _utilEnGarde.eg.string
});

var ZoneVersion = _utilEnGarde.eg.object({
  version: _utilEnGarde.eg.number,
  id: _utilEnGarde.eg.string,
  cloned_from: _utilEnGarde.eg.string,
  application_id: _utilEnGarde.eg.string,
  description: _utilEnGarde.eg.string,
  status: _utilEnGarde.eg.string,
  locked: _utilEnGarde.eg.boolean,
  last_updated: _utilEnGarde.eg.string
});

var CreateZoneVersion = _utilEnGarde.eg.object({
  version: _utilEnGarde.eg.number,
  id: _utilEnGarde.eg.string,
  clone_from: _utilEnGarde.eg.string,
  application_id: _utilEnGarde.eg.string,
  description: _utilEnGarde.eg.string,
  status: _utilEnGarde.eg.string,
  locked: _utilEnGarde.eg.boolean,
  last_updated: _utilEnGarde.eg.string
});

var ZoneInfo = _utilEnGarde.eg.object({
  zone_name: _utilEnGarde.eg.string,
  zone_tag: _utilEnGarde.eg.string
});

exports.ZoneInfo = ZoneInfo;

var WAFMigrationStatus = _utilEnGarde.eg.object({
  new_waf_status: _utilEnGarde.eg.string,
  old_waf_status: _utilEnGarde.eg.string
});
