// @cloudflare/component-header/es/Hamburger.js
import React from 'react';
import PropTypes from 'prop-types';
import { Icon } from '@cloudflare/component-icon';
import { createComponent } from '@cloudflare/style-container';
var Wrapper = createComponent(function (_ref) {
  var theme = _ref.theme;
  return {
    height: '100%',
    display: 'flex',
    alignItems: 'center',
    borderRight: "1px solid ".concat(theme.colors.gray[7]),
    padding: theme.space[3],
    '-webkit-tap-highlight-color': 'rgba(0, 0, 0, 0)',
    userSelect: 'none',
    cursor: 'pointer',
    tabletWide: {
      display: 'none'
    }
  };
}, 'div', ['onClick']);
Wrapper.setDisplayName('Wrapper');

var Hamburger = function Hamburger(_ref2) {
  var onClick = _ref2.onClick,
      label = _ref2.label;
  return React.createElement(Wrapper, {
    onClick: onClick
  }, React.createElement(Icon, {
    label: label,
    size: "1.5x",
    type: "hamburger"
  }));
};

Hamburger.propTypes = {
  label: PropTypes.string,
  onClick: PropTypes.func.isRequired
};
Hamburger.defaultProps = {
  label: 'Open the menu'
};
export default Hamburger;
