// @cloudflare/style-provider/lib/createRenderer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fela = require("fela");

var _felaPluginPrefixer = _interopRequireDefault(require("fela-plugin-prefixer"));

var _felaPluginFallbackValue = _interopRequireDefault(require("fela-plugin-fallback-value"));

var _felaPluginUnit = _interopRequireDefault(require("fela-plugin-unit"));

var _felaPluginEmbedded = _interopRequireDefault(require("fela-plugin-embedded"));

var _felaPluginNamedKeys = _interopRequireDefault(require("fela-plugin-named-keys"));

var _styleConst = require("@cloudflare/style-const");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-line behance/no-deprecated
// eslint-disable-line behance/no-deprecated
var defaultOpts = {
  dev: false
};
var mediaQueries = {
  mobile: "@media (min-width: ".concat(_styleConst.variables.breakpoints.mobile, ")"),
  mobileWide: "@media (min-width: ".concat(_styleConst.variables.breakpoints.mobileWide, ")"),
  mobileOnly: "@media (min-width: ".concat(_styleConst.variables.breakpoints.mobile, ") and (max-width: ").concat(_styleConst.variables.breakpoints.tablet, ")"),
  tablet: "@media (min-width: ".concat(_styleConst.variables.breakpoints.tablet, ")"),
  tabletLegacy: "@media (min-width: ".concat(_styleConst.variables.breakpoints.tabletLegacy, ")"),
  tabletWide: "@media (min-width: ".concat(_styleConst.variables.breakpoints.tabletWide, ")"),
  tabletOnly: "@media (min-width: ".concat(_styleConst.variables.breakpoints.tablet, ") and (max-width: ").concat(_styleConst.variables.breakpoints.desktop, ")"),
  desktop: "@media (min-width: ".concat(_styleConst.variables.breakpoints.desktop, ")"),
  desktopLegacy: "@media (min-width: ".concat(_styleConst.variables.breakpoints.desktopLegacy, ")"),
  desktopLarge: "@media (min-width: ".concat(_styleConst.variables.breakpoints.desktopLarge, ")"),
  hover: "@media (hover)"
};

var removePrefix = function removePrefix(query) {
  return query.replace('@media ', '');
};

var createRenderer = function createRenderer(opts) {
  var usedOpts = Object.assign({}, defaultOpts, opts);
  var plugins = [(0, _felaPluginPrefixer.default)(), (0, _felaPluginFallbackValue.default)(), (0, _felaPluginUnit.default)(), (0, _felaPluginEmbedded.default)(), (0, _felaPluginNamedKeys.default)(mediaQueries)];
  var enhancers = [];

  if (process.env.NODE_ENV === 'development' && usedOpts.dev === true) {
    var validator = require('fela-plugin-validator').default;

    plugins.push(validator());
  }

  return (0, _fela.createRenderer)({
    plugins: plugins,
    enhancers: enhancers,
    selectorPrefix: usedOpts.selectorPrefix,
    devMode: process.env.NODE_ENV === 'development' && usedOpts.dev,
    mediaQueryOrder: [removePrefix(mediaQueries.hover), removePrefix(mediaQueries.mobile), removePrefix(mediaQueries.mobileWide), removePrefix(mediaQueries.tablet), removePrefix(mediaQueries.tabletLegacy), removePrefix(mediaQueries.tabletWide), removePrefix(mediaQueries.desktop), removePrefix(mediaQueries.desktopLegacy), removePrefix(mediaQueries.desktopLarge)]
  });
};

var _default = createRenderer;
exports.default = _default;
