// @cloudflare/style-provider/lib/felaSnapshot.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactTestRenderer = _interopRequireDefault(require("react-test-renderer"));

var _felaTools = require("fela-tools");

var _felaTestContext = _interopRequireDefault(require("./felaTestContext"));

var _createRenderer = _interopRequireDefault(require("./createRenderer"));

var _cssbeautify = _interopRequireDefault(require("cssbeautify"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var options = {
  indent: '  ',
  openbrace: 'end-of-line',
  autosemicolon: false
};

var _default = function _default(component) {
  var renderOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var felaRenderer = (0, _createRenderer.default)({
    dev: true
  });
  return {
    component: _reactTestRenderer.default.create((0, _felaTestContext.default)(component, felaRenderer), renderOptions).toJSON(),
    styles: "\n".concat((0, _cssbeautify.default)((0, _felaTools.renderToString)(felaRenderer), options), "\n")
  };
};

exports.default = _default;
