// @cloudflare/component-label/lib/Label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var typeColors = {
  default: 'gray.4',
  info: 'blue.4',
  success: 'green.4',
  warning: 'orange.5',
  error: 'red.4'
};

var styles = function styles(_ref) {
  var theme = _ref.theme,
      type = _ref.type;
  return {
    alignItems: 'center',
    borderRadius: theme.radii[2],
    color: theme.colors.white,
    display: 'inline-flex',
    fontSize: 1,
    fontWeight: 600,
    lineHeight: theme.lineHeights.solid,
    userSelect: 'none',
    whiteSpace: 'nowrap',
    px: 2,
    py: 1,
    backgroundColor: typeColors[type]
  };
};

var Label = (0, _styleContainer.createStyledComponent)(styles, 'span');
Label.propTypes = {
  type: _propTypes.default.oneOf(['default', 'info', 'success', 'warning', 'error']).isRequired
};
Label.displayName = 'Label';
var _default = Label;
exports.default = _default;
