// @cloudflare/component-input/src/Input.js
import React from 'react';
import PropTypes from 'prop-types';
import { createStyledComponent } from '@cloudflare/style-container';

const styles = ({ theme, loading }) => ({
  mx: 0,
  mt: 0,
  mb: 3,
  p: 2,

  borderStyle: 'solid',
  borderWidth: '1px',
  borderColor: 'gray.5',
  borderRadius: '1px',

  height: 'initial',
  lineHeight: 'initial',

  flex: 'initial',
  zIndex: 0,
  width: '100%',

  verticalAlign: 'middle',
  fontFamily: theme.fontFamily,
  fontSize: 2,
  opacity: loading ? 0.5 : 1,
  backgroundColor: 'white',
  color: 'gray.1',
  outline: 'none',

  transition: 'border-color 0.2s ease',
  textIndent: 'initial',

  '&:hover': {
    borderColor: theme.colors.blue[4]
  },

  '&:disabled': {
    background: 'white',
    color: theme.colors.gray[4],
    '&:hover': {
      borderColor: theme.colors.gray[5]
    }
  }
});

class Input extends React.Component {
  render() {
    const { inputRef, loading, testId, ...restProps } = this.props;

    return <input data-test-id={testId} {...restProps} ref={inputRef} />;
  }
}

Input.propTypes = {
  type: PropTypes.oneOf([
    'text',
    'email',
    'number',
    'password',
    'search',
    'color'
  ]),
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  defaultValue: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func,
  inputRef: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  placeholder: PropTypes.string,
  autoComplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  testId: PropTypes.string
};

Input.defaultProps = {
  type: 'text'
};
Input.displayName = 'Input';

export default createStyledComponent(styles, Input);
