// @cloudflare/component-input/src/LabeledInput.js
import React from 'react';
import PropTypes from 'prop-types';
import { createStyledComponent } from '@cloudflare/style-container';
import { Box } from '@cloudflare/component-box';

import Input from './Input';

const InputLabel = createStyledComponent(({ theme }) => ({
  color: 'gray.3',
  textAlign: 'center',
  backgroundColor: 'gray.9',
  borderColor: 'gray.5',
  borderTop: `1px solid ${theme.colors.gray[5]}`,
  borderBottom: `1px solid ${theme.colors.gray[5]}`,
  borderLeft: `1px solid ${theme.colors.gray[5]}`,
  whiteSpace: 'nowrap',
  marginRight: '-1px',
  borderTopLeftRadius: '1px',
  borderBottomLeftRadius: '1px',
  p: 2,
  fontSize: 2,
  lineHeight: 'initial'
}));
InputLabel.setDisplayName('InputLabel');

const LabeledInput = ({ label, ...otherProps }) => (
  <Box display="flex" width="100%" mb={3}>
    {label && <InputLabel>{label}</InputLabel>}
    <Box flexGrow={1}>
      <Input {...otherProps} mb={0} borderRadius="0 1px 1px 0" />
    </Box>
  </Box>
);
LabeledInput.displayName = 'LabeledInput';

LabeledInput.propTypes = {
  label: PropTypes.string
};

export default LabeledInput;
