// @cloudflare/component-arrow-swivel/es/ArrowSwivel.js
import * as React from 'react';
import PropTypes from 'prop-types';
import { Icon } from '@cloudflare/component-icon';
import { createComponent } from '@cloudflare/style-container';
const CaretIconWrapper = createComponent(({
  isActive,
  startAngle,
  endAngle,
  duration
}) => ({
  position: 'relative',
  top: '0.2rem',
  display: 'inline-block',
  '& svg': {
    transform: `rotate(${isActive ? endAngle : startAngle}deg)`,
    transition: `transform ${duration}ms ease`
  }
}), 'span');

const ArrowSwivel = ({
  onClick,
  isActive = false,
  startAngle = 0,
  endAngle = 90,
  duration = 200,
  color = 'blue.4',
  size
}) => /*#__PURE__*/React.createElement(CaretIconWrapper, {
  startAngle: startAngle,
  endAngle: endAngle,
  onClick: onClick,
  isActive: isActive,
  duration: duration
}, /*#__PURE__*/React.createElement(Icon, {
  label: "caret-right",
  type: "caret-right",
  color: color,
  size: size
}));

ArrowSwivel.propTypes = {
  isActive: PropTypes.bool,
  startAngle: PropTypes.number,
  endAngle: PropTypes.number,
  onClick: PropTypes.func,
  duration: PropTypes.number,
  color: PropTypes.string,
  size: PropTypes.string
};
ArrowSwivel.defaultProps = {
  isActive: false,
  startAngle: 0,
  endAngle: 90,
  duration: 200,
  color: 'blue.4'
};
ArrowSwivel.displayName = 'ArrowSwivel';
export default ArrowSwivel;
