// @cloudflare/style-nextjs/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStyles = exports.getInitialProps = exports.StyleProvider = void 0;

var _react = _interopRequireDefault(require("react"));

var _styleProvider = require("@cloudflare/style-provider");

var _felaDom = require("fela-dom");

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var felaRenderer = (0, _styleProvider.createRenderer)();

var getInitialProps = function getInitialProps(callback) {
  return function (nextObj) {
    var page = nextObj.renderPage();
    var cfFelaSheetList = (0, _felaDom.renderToSheetList)(felaRenderer);
    felaRenderer.clear();
    var userProps = callback ? callback(nextObj) : {};
    return _objectSpread(_objectSpread(_objectSpread({}, page), userProps), {}, {
      cfFelaSheetList: cfFelaSheetList
    });
  };
};

exports.getInitialProps = getInitialProps;

var getStyles = function getStyles(props) {
  return props.cfFelaSheetList.map(function (_ref) {
    var type = _ref.type,
        media = _ref.media,
        css = _ref.css;
    return /*#__PURE__*/_react.default.createElement("style", {
      dangerouslySetInnerHTML: {
        __html: css
      },
      "data-fela-type": type,
      key: "".concat(type, "-").concat(media),
      media: media
    });
  });
};

exports.getStyles = getStyles;

var StyleProvider = function StyleProvider(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/_react.default.createElement(_styleProvider.StyleProvider, {
    renderer: felaRenderer
  }, children);
};

exports.StyleProvider = StyleProvider;
StyleProvider.propTypes = {
  children: _propTypes.default.node
};
