// @cloudflare/component-input/lib/LabeledInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styleContainer = require("@cloudflare/style-container");

var _componentBox = require("@cloudflare/component-box");

var _Input = _interopRequireDefault(require("./Input"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var InputLabel = (0, _styleContainer.createStyledComponent)(function (_ref) {
  var theme = _ref.theme;
  return {
    color: 'gray.3',
    textAlign: 'center',
    backgroundColor: 'gray.9',
    borderColor: 'gray.5',
    borderTop: "1px solid ".concat(theme.colors.gray[5]),
    borderBottom: "1px solid ".concat(theme.colors.gray[5]),
    borderLeft: "1px solid ".concat(theme.colors.gray[5]),
    whiteSpace: 'nowrap',
    marginRight: '-1px',
    borderTopLeftRadius: '1px',
    borderBottomLeftRadius: '1px',
    p: 2,
    fontSize: 2,
    lineHeight: 'initial'
  };
});
InputLabel.setDisplayName('InputLabel');

var LabeledInput = function LabeledInput(_ref2) {
  var label = _ref2.label,
      otherProps = _objectWithoutProperties(_ref2, ["label"]);

  return _react.default.createElement(_componentBox.Box, {
    display: "flex",
    width: "100%",
    mb: 3
  }, label && _react.default.createElement(InputLabel, null, label), _react.default.createElement(_componentBox.Box, {
    flexGrow: 1
  }, _react.default.createElement(_Input.default, _extends({}, otherProps, {
    mb: 0,
    borderRadius: "0 1px 1px 0"
  }))));
};

LabeledInput.displayName = 'LabeledInput';
LabeledInput.propTypes = {
  label: _propTypes.default.string
};
var _default = LabeledInput;
exports.default = _default;
