// cf-style-provider/lib/felaSnapshot.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _reactTestRenderer = require('react-test-renderer');

var _reactTestRenderer2 = _interopRequireDefault(_reactTestRenderer);

var _felaTools = require('fela-tools');

var _felaTestContext = require('./felaTestContext');

var _felaTestContext2 = _interopRequireDefault(_felaTestContext);

var _createRenderer = require('./createRenderer');

var _createRenderer2 = _interopRequireDefault(_createRenderer);

var _cssbeautify = require('cssbeautify');

var _cssbeautify2 = _interopRequireDefault(_cssbeautify);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const options = {
  indent: '  ',
  openbrace: 'end-of-line',
  autosemicolon: false
};

exports.default = component => {
  const felaRenderer = (0, _createRenderer2.default)({
    dev: true
  });

  return {
    component: _reactTestRenderer2.default.create((0, _felaTestContext2.default)(component, felaRenderer)).toJSON(),
    styles: `\n${(0, _cssbeautify2.default)((0, _felaTools.renderToString)(felaRenderer), options)}\n`
  };
};
