// @cloudflare/style-const/lib/colors/darkScalesV2.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = {
  white: '#000',
  black: '#fff',
  red: ['#ffe4e2', '#febcb7', '#fd9087', '#fd5548', '#eb1403', '#cc1203', '#b51002', '#880c02', '#600801', '#3b0501'],
  orange: ['#fde5d0', '#fac190', '#f79847', '#e46e0a', '#bc5b08', '#a34f07', '#904506', '#6b3304', '#4a2403', '#2a1502'],
  gold: ['#f7e9c3', '#ebc96d', '#dfa50c', '#ba8700', '#996f00', '#846000', '#745500', '#563e00', '#3b2b00', '#221900'],
  green: ['#cff3dc', '#81dfa3', '#32c769', '#29a456', '#228747', '#1e753e', '#1a6736', '#134c28', '#0d341b', '#081e10'],
  cyan: ['#d0eff7', '#8dd7eb', '#43bddd', '#229cbd', '#1c809c', '#186f87', '#156277', '#104858', '#0b323c', '#061c22'],
  blue: ['#dcebff', '#acceff', '#79b1ff', '#3e8eff', '#056dff', '#005de0', '#0052c6', '#003e93', '#002b66', '#00193d'],
  indigo: ['#ede8fc', '#d1c5f7', '#b4a4f1', '#9281ea', '#7165e0', '#6357c0', '#584da8', '#423979', '#2e284f', '#1b1729'],
  violet: ['#f5e5fb', '#e7bff4', '#d896ee', '#c768e6', '#b639de', '#a222cc', '#901eb5', '#6b1687', '#4b0f5e', '#2d0939'],
  pink: ['#fee3ec', '#fcbad0', '#fb8db2', '#f8528a', '#e90954', '#cb0849', '#b30741', '#870531', '#5f0423', '#3b0216'],
  gray: ['#eaeaea', '#cccccc', '#aeaeae', '#909090', '#767676', '#666666', '#5a5a5a', '#424242', '#2c2c2c', '#191919'],
  chartreuse: ['#DEF0CF', '#ADD986', '#7AC13C', '#649F32', '#528329', '#477123', '#3F641F', '#2E4917', '#203310', '#121D09'],
  fuchsia: ['#FFE1F5', '#FEB6E7', '#FD85D7', '#FC3DBF', '#DF0399', '#C20385', '#AC0376', '#810259', '#5C013F', '#380127'],
  categorical: ['#ACCEFF', '#186F87', '#FAC190', '#A222CC', '#ADD986', '#C20385', '#FEBCB7', '#6357C0', '#81DFA3', '#CB0849', '#EBC96D', '#477123', '#FCBAD0', '#005DE0', '#FCBAD0', '#1E753E', '#E7BFF4', '#CC1203']
};
exports.default = _default;
