// @cloudflare/style-const/lib/colors/lightScalesV1.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = {
  white: '#fff',
  black: '#000',
  red: ['#3c0501', '#5a0801', '#780a02', '#970d02', '#b20f03', '#e81403', '#fc574a', '#fe9f97', '#feccc8', '#ffefee'],
  orange: ['#361a02', '#482303', '#592b04', '#763905', '#8d4406', '#c05d08', '#ee730a', '#f8a054', '#fbcda5', '#fef1e6'],
  gold: ['#261c00', '#3e2d00', '#4c3700', '#644900', '#735400', '#a77a00', '#dda100', '#ffce4b', '#ffeab2', '#fff8e4'],
  green: ['#0a2614', '#0e381d', '#104122', '#15562d', '#196535', '#228b49', '#2db35e', '#55d584', '#a8e9c0', '#e3f8eb'],
  cyan: ['#061b20', '#0b333e', '#0d3e4b', '#115061', '#156074', '#1e89a5', '#30b6da', '#73cee6', '#ace2f0', '#e9f7fb'],
  blue: ['#001c43', '#002b67', '#003681', '#0045a6', '#0051c3', '#086fff', '#4693ff', '#82b6ff', '#b9d6ff', '#ecf4ff'],
  indigo: ['#170f58', '#221785', '#2c1ea9', '#3524cd', '#4b3bdc', '#7366e4', '#9d94ec', '#c2bdf3', '#dfdcf9', '#f1f0fc'],
  violet: ['#350b42', '#490f5c', '#5f1477', '#741892', '#8d1eb1', '#b73cdf', '#cf7ee9', '#dfa8f1', '#ebcaf6', '#f7eafb'],
  pink: ['#2d0210', '#4e031c', '#6a0426', '#8d0633', '#af0740', '#e80954', '#f85189', '#fb97b9', '#fdc9db', '#fef1f5'],
  gray: ['#1d1d1d', '#313131', '#3d3d3d', '#4a4a4a', '#595959', '#797979', '#999999', '#b6b6b6', '#d9d9d9', '#f2f2f2'],
  categorical: ['#4693FF', '#82B6FF', '#FBAD41', '#FCC373', '#55D584', '#7EDFA2', '#F85189', '#FB97B9', '#B73CDF', '#CF7EE9', '#FFE43E', '#FFEAB2', '#73CEE6', '#ACE2F0', '#E81403', '#FC574A'],
  newGray: ['#191919', '#2C2C2C', '#424242', '#5A5A5A', '#666666', '#767676', '#909090', '#AEAEAE', '#CCCCCC', '#EAEAEA'],
  newGreen: ['#081E10', '#0D341B', '#134C28', '#1A6736', '#1E753E', '#228747', '#29A456', '#32C769', '#81DFA3', '#CFF3DC'],
  sequential: ['#3E8EFF', '#104858', '#E46E0A', '#6B1687', '#F8528A', '#003E93', '#FD5548', '#423979', '#29A456', '#870531', '#BA8700', '#134C28', '#C768E6', '#880C02']
};
exports.default = _default;
