// fela-dom/lib/server/renderToSheetList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderToSheetList;

var _felaUtils = require('fela-utils');

var _getRehydrationIndex = require('./getRehydrationIndex');

var _getRehydrationIndex2 = _interopRequireDefault(_getRehydrationIndex);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sheetMap = {
  fontFaces: _felaUtils.FONT_TYPE,
  statics: _felaUtils.STATIC_TYPE,
  keyframes: _felaUtils.KEYFRAME_TYPE,
  rules: _felaUtils.RULE_TYPE
};

function renderToSheetList(renderer) {
  var cacheCluster = (0, _felaUtils.clusterCache)(renderer.cache, renderer.mediaQueryOrder);

  var rehydrationIndex = (0, _getRehydrationIndex2.default)(renderer);

  var sheetList = (0, _felaUtils.objectReduce)(sheetMap, function (list, type, key) {
    if (cacheCluster[key].length > 0) {
      list.push({
        css: cacheCluster[key],
        rehydration: rehydrationIndex,
        type: type
      });
    }

    return list;
  }, []);

  return (0, _felaUtils.objectReduce)(cacheCluster.mediaRules, function (list, css, media) {
    if (css.length > 0) {
      list.push({
        css: css,
        type: _felaUtils.RULE_TYPE,
        rehydration: rehydrationIndex,
        media: media
      });
    }

    return list;
  }, sheetList);
}
