// @cloudflare/style-container/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  withTheme: true,
  combineRules: true,
  ThemeProvider: true,
  connect: true,
  FelaTheme: true,
  applyStaticStyles: true,
  applyTheme: true,
  createComponent: true,
  createComponentStyles: true,
  createStyledComponent: true,
  filterStyleProps: true,
  withRenderer: true,
  ariaAttributes: true,
  isValidElementProp: true,
  filterValidElementProps: true
};
Object.defineProperty(exports, "FelaTheme", {
  enumerable: true,
  get: function get() {
    return _reactFela.FelaTheme;
  }
});
Object.defineProperty(exports, "ThemeProvider", {
  enumerable: true,
  get: function get() {
    return _reactFela.ThemeProvider;
  }
});
Object.defineProperty(exports, "applyStaticStyles", {
  enumerable: true,
  get: function get() {
    return _applyStaticStyles.default;
  }
});
Object.defineProperty(exports, "applyTheme", {
  enumerable: true,
  get: function get() {
    return _applyTheme.default;
  }
});
Object.defineProperty(exports, "ariaAttributes", {
  enumerable: true,
  get: function get() {
    return _isValidElementProp.ariaAttributes;
  }
});
Object.defineProperty(exports, "combineRules", {
  enumerable: true,
  get: function get() {
    return _fela.combineRules;
  }
});
Object.defineProperty(exports, "connect", {
  enumerable: true,
  get: function get() {
    return _reactFela.connect;
  }
});
Object.defineProperty(exports, "createComponent", {
  enumerable: true,
  get: function get() {
    return _createComponent.default;
  }
});
Object.defineProperty(exports, "createComponentStyles", {
  enumerable: true,
  get: function get() {
    return _createComponentStyles.default;
  }
});
Object.defineProperty(exports, "createStyledComponent", {
  enumerable: true,
  get: function get() {
    return _createStyledComponent.default;
  }
});
Object.defineProperty(exports, "filterStyleProps", {
  enumerable: true,
  get: function get() {
    return _createStyledComponent.filterStyleProps;
  }
});
Object.defineProperty(exports, "filterValidElementProps", {
  enumerable: true,
  get: function get() {
    return _isValidElementProp.filterValidElementProps;
  }
});
Object.defineProperty(exports, "isValidElementProp", {
  enumerable: true,
  get: function get() {
    return _isValidElementProp.isValidElementProp;
  }
});
Object.defineProperty(exports, "withRenderer", {
  enumerable: true,
  get: function get() {
    return _withRenderer.default;
  }
});
exports.withTheme = void 0;

var _fela = require("fela");

var _reactFela = require("react-fela");

var _applyStaticStyles = _interopRequireDefault(require("./applyStaticStyles"));

var _applyTheme = _interopRequireDefault(require("./applyTheme"));

var _createComponent = _interopRequireDefault(require("./createComponent"));

var _createComponentStyles = _interopRequireDefault(require("./createComponentStyles"));

var _createStyledComponent = _interopRequireWildcard(require("./createStyledComponent"));

var _withRenderer = _interopRequireDefault(require("./withRenderer"));

var _isValidElementProp = require("./isValidElementProp");

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-line behance/no-deprecated
// eslint-disable-line behance/no-deprecated
var withTheme = _reactFela.withTheme;
exports.withTheme = withTheme;
